cpp/unit/producer.cpp

```diff
@@ -519,49 +521,51 @@ void BuldingProducer::initialise(Unit *unit, Player &player) {
 	unit->give_ability(std::make_shared<UngarrisonAbility>());
 }
 
-std::shared_ptr<TerrainObject> BuldingProducer::place(Unit *u, Terrain &terrain, coord::phys3 init_pos) const {
-
-	// not a tc --- hacks / fix me
-	bool collisions = this->unit_data.id0 != 109;
+std::shared_ptr<TerrainObject> BuldingProducer::place(Unit *u, std::shared_ptr<Terrain> terrain, coord::phys3 init_pos) const {
 
 	// buildings have a square base
-	auto obj = std::make_shared<SquareObject>(*u, this->foundation_size, this->terrain_outline, collisions);
+	u->make_location<SquareObject>(this->foundation_size, this->terrain_outline);
 
 	/*
 	 * decide what terrain is passable using this lambda
 	 * currently unit avoids water and tiles with another unit
 	 * this function should be true if pos is a valid position of the object
 	 */
-	obj->passable = [obj, &terrain](const coord::phys3 &pos) -> bool {
+	std::weak_ptr<TerrainObject> obj_ptr = u->location;
+	std::weak_ptr<Terrain> terrain_ptr = terrain;
+	u->location->passable = [obj_ptr, terrain_ptr](const coord::phys3 &pos) -> bool {
+		auto obj = obj_ptr.lock();
+		auto terrain = terrain_ptr.lock();
 
 		// look at all tiles in the bases range
 		for (coord::tile check_pos : tile_list(obj->get_range(pos))) {
-			TileContent *tc = terrain.get_data(check_pos);
+			TileContent *tc = terrain->get_data(check_pos);
 			if (!tc) {
 				return false;
 			}
 			for (auto item : tc->obj) {
 				auto tobj = item.lock();
-				if (tobj->check_collisions) return false;
+				if (tobj->check_collisions()) return false;
 			}
 		}
 		return true;
 	};
 
 	// drawing function
-	obj->draw = [=]() {
+	// not a tc --- hacks / fix me
+	bool collisions = this->unit_data.id0 != 109;
+	u->location->draw = [u, obj_ptr, collisions]() {
 		if (u->selected && collisions) {
-			obj->draw_outline();
+			obj_ptr.lock()->draw_outline();
 		}
 		u->draw();
 	};
 
 	// try to place the obj, it knows best whether it will fit.
-	bool obj_placed = obj->place(&terrain, init_pos);
-	if (!obj_placed) {
+	auto state = object_state::floating; //collisions? object_state::placed : object_state::placed_no_collision;
```

Comment:
Its not mine, I kept it from the original placement function
It means place() does the checks as to whether placement is allowed